Vagrant post-processor literally has "panic" placeholder in Packer 1.5.x in HCL2 mode

#### Overview of the Issue

PR #8423 breaks Vagrant post-processor with panic instruction. Haven't seen other issue tracking this so I thought it's worth to open a separate one.

https://github.com/hashicorp/packer/blob/c3c2622204fbc4358f1761d38ba99616bf1eba33/post-processor/vagrant/post-processor.go#L63

#### Reproduction Steps

Run following buildfile with .pkr.hcl extension, e.g. `packer build file.pkr.hcl`

### Packer version

1.5.1

### Simplified Packer Buildfile

```
source "virtualbox-iso" "packer-debian-10-amd64" {
  iso_urls = [
    "https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/debian-10.2.0-amd64-netinst.iso"
  ]
  iso_checksum = "e43fef979352df15056ac512ad96a07b515cb8789bf0bfd86f99ed0404f885f5"
  ssh_username = "vagrant"
}

build {
  sources = [
    "source.virtualbox-iso.packer-debian-10-amd64"
  ]

  post-processor "vagrant" {
    output = "ephemeral/builds/vagrant-debian10.box"
  }
}
```

### Operating system and Environment details

OS X Catalina

### Log Fragments and crash.log files

```
$ packer build base_debian_vagrant.pkr.hcl
panic: not implemented yet
2020/01/08 15:47:57 packer-post-processor-vagrant plugin:
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: goroutine 71 [running]:
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: github.com/hashicorp/packer/post-processor/vagrant.(*PostProcessor).ConfigSpec(0xc00015e010, 0xc00001a630)
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: 	/private/tmp/packer-20191222-94130-1yh62h9/post-processor/vagrant/post-processor.go:63 +0x39
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: github.com/hashicorp/packer/packer/rpc.(*commonServer).ConfigSpec(0xc0000aba98, 0x0, 0x0, 0xc0002e6ec0, 0x0, 0x0)
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: 	/private/tmp/packer-20191222-94130-1yh62h9/packer/rpc/common.go:56 +0x38
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: reflect.Value.call(0xc0005dc5a0, 0xc000010768, 0x13, 0x7c7258a, 0x4, 0xc000093f18, 0x3, 0x3, 0x0, 0x0, ...)
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: 	/usr/local/Cellar/go/1.13.5/libexec/src/reflect/value.go:460 +0x5f6
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: reflect.Value.Call(0xc0005dc5a0, 0xc000010768, 0x13, 0xc000080f18, 0x3, 0x3, 0x0, 0x0, 0x0)
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: 	/usr/local/Cellar/go/1.13.5/libexec/src/reflect/value.go:321 +0xb4
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: net/rpc.(*service).call(0xc0000abac0, 0xc0000ee820, 0xc0005b03c0, 0xc0005b03d0, 0xc000136380, 0xc0002e6e20, 0x71f50e0, 0xc0002aa3e0, 0x194, 0x6ef1da0, ...)
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: 	/usr/local/Cellar/go/1.13.5/libexec/src/net/rpc/server.go:377 +0x16f
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: created by net/rpc.(*Server).ServeCodec
2020/01/08 15:47:57 packer-post-processor-vagrant plugin: 	/usr/local/Cellar/go/1.13.5/libexec/src/net/rpc/server.go:474 +0x42b
2020/01/08 15:47:57 ConfigSpec failed: unexpected EOF
2020/01/08 15:47:57 waiting for all plugin processes to complete...
2020/01/08 15:47:57 /usr/local/bin/packer: plugin process exited
2020/01/08 15:47:57 /usr/local/bin/packer: plugin process exited
panic: ConfigSpec failed: unexpected EOF [recovered]
	panic: ConfigSpec failed: unexpected EOF
```

--- a/post-processor/vagrant/post-processor.go
+++ b/post-processor/vagrant/post-processor.go
@@ -56,44 +56,27 @@ type Config struct {
 }
 
 type PostProcessor struct {
-	configs map[string]*Config
+	config Config
 }
 
 func (p *PostProcessor) ConfigSpec() hcldec.ObjectSpec {
-	panic("not implemented yet")
-	// return p.config.FlatMapstructure().HCL2Spec()
+	return p.config.FlatMapstructure().HCL2Spec()
 }
 
 func (p *PostProcessor) Configure(raws ...interface{}) error {
-	p.configs = make(map[string]*Config)
-	p.configs[""] = new(Config)
-	if err := p.configureSingle(p.configs[""], raws...); err != nil {
+	if err := p.configureSingle(&p.config, raws...); err != nil {
 		return err
 	}
-
-	// Go over any of the provider-specific overrides and load those up.
-	for name, override := range p.configs[""].Override {
-		subRaws := make([]interface{}, len(raws)+1)
-		copy(subRaws, raws)
-		subRaws[len(raws)] = override
-
-		config := new(Config)
-		p.configs[name] = config
-		if err := p.configureSingle(config, subRaws...); err != nil {
-			return fmt.Errorf("Error configuring %s: %s", name, err)
-		}
-	}
-
 	return nil
 }
 
 func (p *PostProcessor) PostProcessProvider(name string, provider Provider, ui packer.Ui, artifact packer.Artifact) (packer.Artifact, bool, error) {
-	config := p.configs[""]
-	if specificConfig, ok := p.configs[name]; ok {
-		config = specificConfig
+	config, err := p.specificConfig(name)
+	if err != nil {
+		return nil, false, err
 	}
 
-	err := CreateDummyBox(ui, config.CompressionLevel)
+	err = CreateDummyBox(ui, config.CompressionLevel)
 	if err != nil {
 		return nil, false, err
 	}
@@ -244,6 +227,17 @@ func (p *PostProcessor) configureSingle(c *Config, raws ...interface{}) error {
 	}
 
 	return nil
+}
+
+func (p *PostProcessor) specificConfig(name string) (Config, error) {
+	config := p.config
+	if _, ok := config.Override[name]; ok {
+		if err := mapstructure.Decode(config.Override[name], &config); err != nil {
+			err = fmt.Errorf("Error overriding config for %s: %s", name, err)
+			return config, err
+		}
+	}
+	return config, nil
 }
 
 func providerForName(name string) Provider {

--- a/post-processor/vagrant/post-processor_test.go
+++ b/post-processor/vagrant/post-processor_test.go
@@ -46,7 +46,7 @@ func TestPostProcessorPrepare_compressionLevel(t *testing.T) {
 		t.Fatalf("err: %s", err)
 	}
 
-	config := p.configs[""]
+	config := p.config
 	if config.CompressionLevel != flate.DefaultCompression {
 		t.Fatalf("bad: %#v", config.CompressionLevel)
 	}
@@ -58,7 +58,7 @@ func TestPostProcessorPrepare_compressionLevel(t *testing.T) {
 		t.Fatalf("err: %s", err)
 	}
 
-	config = p.configs[""]
+	config = p.config
 	if config.CompressionLevel != 7 {
 		t.Fatalf("bad: %#v", config.CompressionLevel)
 	}
@@ -83,43 +83,48 @@ func TestPostProcessorPrepare_outputPath(t *testing.T) {
 	}
 }
 
-func TestPostProcessorPrepare_subConfigs(t *testing.T) {
+func TestSpecificConfig(t *testing.T) {
 	var p PostProcessor
-
-	f, err := ioutil.TempFile("", "packer")
-	if err != nil {
-		t.Fatalf("err: %s", err)
-	}
-	defer os.Remove(f.Name())
 
 	// Default
 	c := testConfig()
-	c["compression_level"] = 42
-	c["vagrantfile_template"] = f.Name()
+	c["compression_level"] = 1
+	c["output"] = "folder"
 	c["override"] = map[string]interface{}{
 		"aws": map[string]interface{}{
 			"compression_level": 7,
 		},
 	}
-	err = p.Configure(c)
+	if err := p.Configure(c); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	// overrides config
+	config, err := p.specificConfig("aws")
 	if err != nil {
 		t.Fatalf("err: %s", err)
 	}
 
-	if p.configs[""].CompressionLevel != 42 {
-		t.Fatalf("bad: %#v", p.configs[""].CompressionLevel)
+	if config.CompressionLevel != 7 {
+		t.Fatalf("bad: %#v", config.CompressionLevel)
 	}
 
-	if p.configs[""].VagrantfileTemplate != f.Name() {
-		t.Fatalf("bad: %#v", p.configs[""].VagrantfileTemplate)
+	if config.OutputPath != "folder" {
+		t.Fatalf("bad: %#v", config.OutputPath)
 	}
 
-	if p.configs["aws"].CompressionLevel != 7 {
-		t.Fatalf("bad: %#v", p.configs["aws"].CompressionLevel)
+	// does NOT overrides config
+	config, err = p.specificConfig("virtualbox")
+	if err != nil {
+		t.Fatalf("err: %s", err)
 	}
 
-	if p.configs["aws"].VagrantfileTemplate != f.Name() {
-		t.Fatalf("bad: %#v", p.configs["aws"].VagrantfileTemplate)
+	if config.CompressionLevel != 1 {
+		t.Fatalf("bad: %#v", config.CompressionLevel)
+	}
+
+	if config.OutputPath != "folder" {
+		t.Fatalf("bad: %#v", config.OutputPath)
 	}
 }
 

